### PR TITLE
Automatically import binary extensions whose suffixes include the python version

### PR DESCRIPTION
**Pull request by ian-hurst**
TODO:

- [ ] Write new tests or update the old ones to cover new functionality.
- [x] Update doc-strings where appropriate.
- [ ] Update or write new documentation in `packit/packit.dev`.

For an edge case where the current interpreter version differs from that of the rpm library, and where that library's extension suffixes are included in the `_rpm.so` filename (e.g. `_rpm.cpython-39-x86_64-linux-gnu.so` on RHEL 9.x), add a new `reload_module` method that finds the matching .so and imports it directly.

The basic operation is:

- gather a list of valid extension suffixes for the rpm library's interpreter, via `importlib.machinery.EXTENSION_SUFFIXES`
- try the import but catch any `ModuleNotFoundError`
- on `ModuleNotFoundError`, loop through the suffix list, looking for a matching `{module}.{suffix}.so` on disk
- when a match is found, import it as `rpm.{module}` directly by path

This is probably naive or unwise in some way, but I wanted to take a stab at it so I can use rpm-shim in on RHEL 9.x systems at my company without needing to use the `sudo ln` workaround

Example from python3.11 on el9 (Oracle 9.x):

```
DEBUG:rpm-shim:Collected sitepackages and extension suffixes for /usr/libexec/platform-python:
{'sitepackages': ['/usr/local/lib64/python3.9/site-packages',
                  '/usr/local/lib/python3.9/site-packages',
                  '/usr/lib64/python3.9/site-packages',
                  '/usr/lib/python3.9/site-packages'],
 'suffixes': ['.cpython-39-x86_64-linux-gnu.so', '.abi3.so', '.so']}
DEBUG:rpm-shim:Collected sitepackages and extension suffixes for /usr/bin/python3:
{'sitepackages': ['/usr/local/lib64/python3.9/site-packages',
                  '/usr/local/lib/python3.9/site-packages',
                  '/usr/lib64/python3.9/site-packages',
                  '/usr/lib/python3.9/site-packages'],
 'suffixes': ['.cpython-39-x86_64-linux-gnu.so', '.abi3.so', '.so']}
DEBUG:rpm-shim:Collected sitepackages and extension suffixes for /usr/bin/python3.11:
{'sitepackages': ['/usr/local/lib64/python3.11/site-packages',
                  '/usr/local/lib/python3.11/site-packages',
                  '/usr/lib64/python3.11/site-packages',
                  '/usr/lib/python3.11/site-packages'],
 'suffixes': ['.cpython-311-x86_64-linux-gnu.so', '.abi3.so', '.so']}
DEBUG:rpm-shim:Trying /usr/local/lib64/python3.9/site-packages
DEBUG:rpm-shim:Trying /usr/local/lib/python3.9/site-packages
DEBUG:rpm-shim:Trying /usr/lib64/python3.9/site-packages
DEBUG:rpm-shim:Module rpm._rpm not found in /usr/lib64/python3.9/site-packages/rpm, looking for alternative filenames trying valid extension suffixes
DEBUG:rpm-shim:Trying to load rpm._rpm from /usr/lib64/python3.9/site-packages/rpm/_rpm.cpython-39-x86_64-linux-gnu.so
DEBUG:rpm-shim:Loaded rpm._rpm from /usr/lib64/python3.9/site-packages/rpm/_rpm.cpython-39-x86_64-linux-gnu.so
DEBUG:rpm-shim:Reloaded rpm
DEBUG:rpm-shim:Import successfull
```

### CODE REVIEW

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]: This could be done in a single call, returning both paths and suffixes. If the returned object is a dict with descriptive keys, you can avoid mixing paths and suffixes together in the log output below and print and return the dict directly.

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]: I don't think this is necessary, it causes the module to be executed for a second time.

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py:185]:
```suggestion
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    sys.modules[__name__] = module
```

The module is loaded by `exec_module()`, it just needs to be referenced from `sys.modules` in place of the original module (the shim). There is no need to call `importlib.reload()` after this.

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]: I don't think this is necessary, we know the name of the module we are trying to import (`rpm._rpm`). Also, the plain `.so` extension is in the list of suffixes so it will be processed anyway.
I would probably merge this method with `try_path()`, as there shouldn't be a lot of code left in either method.

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
            if not so.is_file():
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py:9]: The import above should have already imported the `util` (sub)module.

**Review by nforro — CHANGES_REQUESTED** [review]
Thanks, this looks interesting!
I left some comments. It would be also nice to have some test cases for this.

File: rpm/__init__.py
Comment (nforro) [reply rpm/__init__.py]: For example: 
```python
def get_sitepackages_and_suffixes(interpreter):
    script = textwrap.dedent(
        """
        import importlib
        import json
        import site

        print(
            json.dumps(
                {
                    "sitepackages": site.getsitepackages(),
                    "suffixes": importlib.machinery.EXTENSION_SUFFIXES,
                }
            )
        )
        """
    )
    output = subprocess.check_output([interpreter], input=script.encode())
    return json.loads(output.decode())
```

**Comment by ian-hurst** [comment]
Thanks for the feedback! Going through it now. I'm happy to add testcases if this is ultimately seen as viable

File: rpm/__init__.py
Comment (ian-hurst) [reply rpm/__init__.py:9]: It (really weirdly) doesn't work unless I import it explicitly:

```
$ python
Python 3.11.7 (main, Aug 23 2024, 00:00:00) [GCC 11.4.1 20231218 (Red Hat 11.4.1-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import importlib
>>> importlib.util
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'importlib' has no attribute 'util'
>>> import importlib.util
>>> importlib.util
<module 'importlib.util' (frozen)>
```

I am totally perplexed by this.

Comment (nforro) [reply rpm/__init__.py:9]: Ah, maybe it's imported in the background by `importlib.reload()` then. Sorry, my bad.

File: rpm/__init__.py
Comment (ian-hurst) [reply rpm/__init__.py]: ok nice, confirmed it's not needed

File: rpm/__init__.py
Comment (ian-hurst) [reply rpm/__init__.py]: What I have to do here is transform `rpm._rpm` from `e.name` to the file path `{sitepackages}/rpm/_rpm.cpython-39-x86_64-linux-gnu.so`. I could just assume that there will always be exactly one extension named `_rpm{suffix}` I guess? It would be safe in python 3.9 on RHEL 9. But for example, on older versions of the python bindings, there are other extensions like `_rpmb` and `_rpms`, and the assumption would fail in any such case. That doesn't seem safe to me. Is there another way to translate {e.name} into an expected extension path?

Comment (ian-hurst) [reply rpm/__init__.py]: as far as merging into try_path, can do, but I trying to avoid a nested try/except, just for legibility, like:

```
try:
    try:
        importlib.reload(sys.modules[__name__])
    except ModuleNotFounError as e:
       {...}
    confdir = 
finally:
```

Comment (nforro) [reply rpm/__init__.py]: > What I have to do here is transform `rpm._rpm` from `e.name` to the file path `{sitepackages}/rpm/_rpm.cpython-39-x86_64-linux-gnu.so`. I could just assume that there will always be exactly one extension named `_rpm{suffix}` I guess? It would be safe in python 3.9 on RHEL 9. But for example, on older versions of the python bindings, there are other extensions like `_rpmb` and `_rpms`, and the assumption would fail in any such case. That doesn't seem safe to me. Is there another way to translate {e.name} into an expected extension path?

Oh, I see, I didn't fully think about what's going on. Importing `rpm` package fails as a result of failure to import the binary module. In that case your logic is correct, I don't think there is any other way but to get the module/file name from the exception. However there could be more binary modules (like you said, older versions had `_rpm`, `_rpmb` and `_rpms`, all imported from `rpm`), so it should be done in a loop. That also means that my `sys.modules[__name__] = module` suggestion is wrong, it has to be `sys.modules[module_name] = module`. And there should be one final `importlib.reload(sys.modules[__name__])` to reload `rpm` once the binary module(s) is/are loaded.

Comment (ian-hurst) [reply rpm/__init__.py]: Ok, makes sense. I will rework this a bit

**Comment by ian-hurst** [comment]
Ok, I reworked this to just loop through `*.cpython*.so` on `ModuleNotFoundError` exception. If the glob matches any binary extensions, I just load them. One benefit of this is I no longer need any of the suffix stuff. Debug output ends up:

```
DEBUG:rpm-shim:Collected sitepackages for /usr/libexec/platform-python:
/usr/local/lib64/python3.9/site-packages
/usr/local/lib/python3.9/site-packages
/usr/lib64/python3.9/site-packages
/usr/lib/python3.9/site-packages
DEBUG:rpm-shim:Collected sitepackages for /usr/bin/python3:
/usr/local/lib64/python3.9/site-packages
/usr/local/lib/python3.9/site-packages
/usr/lib64/python3.9/site-packages
/usr/lib/python3.9/site-packages
DEBUG:rpm-shim:Collected sitepackages for /usr/bin/python3.11:
/usr/local/lib64/python3.11/site-packages
/usr/local/lib/python3.11/site-packages
/usr/lib64/python3.11/site-packages
/usr/lib/python3.11/site-packages
DEBUG:rpm-shim:Trying /usr/lib/python3.11/site-packages
DEBUG:rpm-shim:Trying /usr/lib64/python3.9/site-packages
DEBUG:rpm-shim:The rpm._rpm module was not found in /usr/lib64/python3.9/site-packages/rpm, try loading binary extensions directly
DEBUG:rpm-shim:Found binary extensions:
[PosixPath('/usr/lib64/python3.9/site-packages/rpm/_rpm.cpython-39-x86_64-linux-gnu.so')]
DEBUG:rpm-shim:importing /usr/lib64/python3.9/site-packages/rpm/_rpm.cpython-39-x86_64-linux-gnu.so as rpm._rpm
DEBUG:rpm-shim:imported extensions, reloading rpm
DEBUG:rpm-shim:Import successfull
```

**Comment by nforro** [comment]
> Ok, I reworked this to just loop through `*.cpython*.so` on `ModuleNotFoundError` exception. If the glob matches any binary extensions, I just load them. One benefit of this is I no longer need any of the suffix stuff.

While this is more straightforward, I think I liked the previous solution better. Doing it this way it could attempt to load unnecessary modules and that could cause problems, I can imagine a situation where in an unversioned sitepackages path (I think Debian/Ubuntu use those) there are binary modules in multiple versions and loading all of them would be impossible. Getting module name from the exception and constructing filename by trying valid suffixes seems like a better way.

**Comment by ian-hurst** [comment]
> While this is more straightforward, I think I liked the previous solution better. Doing it this way it could attempt to load unnecessary modules and that could cause problems, I can imagine a situation where in an unversioned sitepackages path (I think Debian/Ubuntu use those) there are binary modules in multiple versions and loading all of them would be impossible. Getting module name from the exception and constructing filename by trying valid suffixes seems like a better way.

yeah, definitely a drawback to this method. The alternative, I think, would be to do something like reload() in a loop, and break when either:

- reload() succeeds, or
- the helper has exhausted all binary extensions, and reload() is still failing
 
If that sounds good (or if there's a better way) I'm happy to go with it.

**Comment by nforro** [comment]
> The alternative, I think, would be to do something like reload() in a loop, and break when either:
> 
> * reload() succeeds, or
> * the helper has exhausted all binary extensions, and reload() is still failing
> 
> If that sounds good (or if there's a better way) I'm happy to go with it.

Yes, that's what I had in mind - loop until there are no more recoverable `ModuleNotFoundError` exceptions or until it's not possible to recover.

**Comment by ian-hurst** [comment]
> Yes, that's what I had in mind - loop until there are no more recoverable `ModuleNotFoundError` exceptions or until it's not possible to recover.

Alright, super! Working on it now

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
def get_system_sitepackages_and_suffixes() -> List[Dict[str, List[str]]]:
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        sitepackages_and_suffixes = get_sitepackages_and_suffixes(interpreter)
        logger.debug(
            f"Collected sitepackages and extension suffixes for {interpreter}:\n"
            + pprint.pformat(sitepackages_and_suffixes)
        )
        result.append(sitepackages_and_suffixes)
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]: Function name should be a verb, not a noun:

```suggestion
def reload_module(path: Path, suffixes: List[str]) -> None:
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    Reloads the `rpm` module. In case some of the required binary submodules fail to import,
    tries to import them directly by path using valid extension suffixes.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]: The types are redundant, they are already present in the function signature.

```suggestion
        path: Absolute path to the `rpm` module.
        suffixes: List of extension suffixes valid for the Python interpreter associated
          with the path.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
                "looking for alternative filenames trying valid extension suffixes"
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
def try_import_binary_extension(path: Path, module: str, suffixes: List[str]) -> bool:
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    Tries to find and import a binary extension in the specified path based on name
    and valid extension suffixes.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        path: Path to the module, e.g. /usr/lib64/python3.9/site-packages/rpm/
        module: Name of the module to import, e.g. `rpm._rpm`
        suffixes: List of extension suffixes to check for
          (see `importlib.machinery.EXTENSION_SUFFIXES`).
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        True if the module was loaded, False otherwise.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    # get the submodule name, e.g. just '_rpm' from 'rpm._rpm'
    submodule = module.rpartition(".")[-1]
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        # a file named {child}{suffix} may exist in {path}, e.g.:
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        logger.debug(f"No combination of {child} and valid suffixes found in {path}, giving up")
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    Imports a Python module by path.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        module_name: Name of the module to import.
        path: Absolute path to the module to import.
        
    Returns:
        True if the import succeeded, otherwise False.
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    logger.debug(f"Trying to load {module_name} from {path}")
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
    for entry in get_system_sitepackages_and_suffixes():
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
        for path in entry["sitepackages"]:
```

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py]:
```suggestion
                if try_path(path, entry["suffixes"]):
```

**Review by nforro — COMMENTED** [review]
These are all style suggestions, the code looks fine :+1:

**Comment by ian-hurst** [comment]
> These are all style suggestions, the code looks fine 👍

Alright, thanks for the feedback! I've taken all the suggested changes now

File: rpm/__init__.py
Comment (nforro) [inline rpm/__init__.py:116]: Ok, the mypy error means that `e.name` in this exception block can be `None`, so there needs to be a check for that.

Comment (ian-hurst) [reply rpm/__init__.py:116]: Yeah I originally had an e.name value check and forgot why. It was for this. Fixed again

**Review by nforro — APPROVED** [review]
Great, thanks. Only some test cases are missing, but I'll try to come up with some, unless you want to do it yourself.

Could you please squash the commits into one?

**Comment by ian-hurst** [comment]
Alright, thanks so much for your review and feedback. I love this little shim and I'm so happy to be able to contribute. As an internet rando, I don't actually have write access, so I can't do the merge step myself.

> Only some test cases are missing, but I'll try to come up with some, unless you want to do it yourself.

I'm not familiar with your test framework, but I can think of a few basic designs. The obvious one is the positive test case, on rhel9, install python3.11, create a virtualenv, then `import rpm`. It might also be good to test that the binary extension import by path works when multiple extensions are used. Can't do that on rhel9 which only has one. But on rhel8 you could create that condition by running `sudo rm -f /usr/lib64/python3.6/site-packages/rpm/_{rpm,rpmb,rpms}.so`. The `.cpython-36m-x86_64-linux-gnu.so` files would remain and then you'd be able to test your loop. I did this manually and it works.

But seems like it might be good to have some negative test cases:

- if _rpm*.so are absent, import should fail
- if _rpm*.so are present, but have invalid suffixes for the original interpreter (probably never occurs in the real world but a good test that things fail)

**Comment by nforro** [comment]
> Alright, thanks so much for your review and feedback. I love this little shim and I'm so happy to be able to contribute. As an internet rando, I don't actually have write access, so I can't do the merge step myself.

I will merge the PR once you squash the commits into one (or more if you feel that's necessary, but I think one is fine). Also, could you please follow the [guidelines](https://github.com/packit/contributing?tab=readme-ov-file#git-workflow) when doing it? :slightly_smiling_face: 

> I'm not familiar with your test framework

There isn't really any framework, just a script that imports `rpm`, either in a virtualenv or directly on a system.

> The obvious one is the positive test case, on rhel9, install python3.11, create a virtualenv, then `import rpm`.

Yes, that's what I had in mind, and it shouldn't be diffcult to modify existing test case to install and run Python 3.11.

> It might also be good to test that the binary extension import by path works when multiple extensions are used. Can't do that on rhel9 which only has one. But on rhel8 you could create that condition by running `sudo rm -f /usr/lib64/python3.6/site-packages/rpm/_{rpm,rpmb,rpms}.so`. The `.cpython-36m-x86_64-linux-gnu.so` files would remain and then you'd be able to test your loop. I did this manually and it works.

Nice idea.

**Comment by ian-hurst** [comment]
> I will merge the PR once you squash the commits into one (or more if you feel that's necessary, but I think one is fine). Also, could you please follow the [guidelines](https://github.com/packit/contributing?tab=readme-ov-file#git-workflow) when doing it? 🙂

sure, will do!
 
> > The obvious one is the positive test case, on rhel9, install python3.11, create a virtualenv, then `import rpm`.
> 
> Yes, that's what I had in mind, and it shouldn't be diffcult to modify existing test case to install and run Python 3.11.

Ok, I'm happy to write tests. Just looking through the workflow now, the same test is run on all flavors of linux. How would you approach adding a test that only runs on one of them?

**Comment by nforro** [comment]
> Ok, I'm happy to write tests.

Thanks!

> Just looking through the workflow now, the same test is run on all flavors of linux. How would you approach adding a test that only runs on one of them?

There could be another environment variable that would be passed to the container and the scripts would react accordingly. Though technically for the simplest case of installing and running in Python 3.11 on RHEL9 (ubi9) it can be simply added to `INSTALL_DEPS_CMD`, tox should run for all interpreters present on the system.

Actually, even the removal of binary modules could be done in `INSTALL_DEPS_CMD`, there is probably no need to edit [tests/entrypoint.sh](https://github.com/packit/rpm-shim/blob/e2c4c95c25e96609a28cb6d0f0f3834f84600430/tests/entrypoint.sh).

**Comment by ian-hurst** [comment]
Ok, I have this squashed and a couple scenarios added. Unfortunately (or really, fortunately) the scenario I set up with python36 reveals an issue. `_rpms` and `_rpmb` aren't being imported. Why not? Because no exception is raised when they fail. Here is `/usr/lib64/python3.6/site-packages/rpm/__init__.py`:

```
# try to import build bits but dont require it
try:
    from rpm._rpmb import *
except ImportError:
    pass

# try to import signing bits but dont require it
try:
    from rpm._rpms import *
except ImportError:
    pass
```

Nothing is raised, so nothing is caught, and then nothing is helped.

**Comment by nforro** [comment]
> Ok, I have this squashed and a couple scenarios added.

Excellent, thanks. The commit message title is a bit long, but that's ok. It shouldn't end with a dot though.

> Unfortunately (or really, fortunately) the scenario I set up with python36 reveals an issue. `_rpms` and `_rpmb` aren't being imported. Why not? Because no exception is raised when they fail. Here is `/usr/lib64/python3.6/site-packages/rpm/__init__.py`:
> 
> ```
> # try to import build bits but dont require it
> try:
>     from rpm._rpmb import *
> except ImportError:
>     pass
> 
> # try to import signing bits but dont require it
> try:
>     from rpm._rpms import *
> except ImportError:
>     pass
> ```
> 
> Nothing is raised, so nothing is caught, and then nothing is helped.

Ok, it's a bit hackish, but I suppose we could do something like `sed -i 's/^    pass/    raise/' /usr/lib64/python3.6/site-packages/rpm/__init__.py` after installation.

**Comment by ian-hurst** [comment]
> Ok, it's a bit hackish, but I suppose we could do something like `sed -i 's/^ pass/ raise/' /usr/lib64/python3.6/site-packages/rpm/__init__.py` after installation.

Yeah, and this is still an improvement, since none of the existing python-rpm versions require this. All old versions include the unversioned .so files, and all new versions have just one extension, _rpm, which works. But still, it's agonizing to be so close to a robust solution and then have to settle. 

One thing I toyed with was modifying `importlib.machinery.EXTENSION_SUFFIXES`. If you make that the union of the list from both interpreters, you can use `pkgutil` to inspect all modules cleanly:

```
>>> importlib.machinery.EXTENSION_SUFFIXES
['.cpython-36m-x86_64-linux-gnu.so', '.cpython-311-x86_64-linux-gnu.so', '.abi3.so', '.so']
>>> [print(i) for i in pkgutil.iter_modules(['/usr/lib64/python3.6/site-packages/rpm'])]
ModuleInfo(module_finder=FileFinder('/usr/lib64/python3.6/site-packages/rpm'), name='_rpm', ispkg=False)
ModuleInfo(module_finder=FileFinder('/usr/lib64/python3.6/site-packages/rpm'), name='_rpmb', ispkg=False)
ModuleInfo(module_finder=FileFinder('/usr/lib64/python3.6/site-packages/rpm'), name='_rpms', ispkg=False)
ModuleInfo(module_finder=FileFinder('/usr/lib64/python3.6/site-packages/rpm'), name='transaction', ispkg=False)
```

So you could use that to load _rpmb and _rpms dynamically. But you would have to assume that all modules should be imported. It's really too bad the actual loader machinery doesn't care what's in EXTENSION_SUFFIXES. If it did, we'd be home free. It's clear python just does not want you to use anything but native suffixes, haha.

**Comment by nforro** [comment]
Ok, let me merge this. Thanks again for your contribution.

**Comment by ian-hurst** [comment]
> Ok, let me merge this. Thanks again for your contribution.

My pleasure, thank you so much for taking this change, and for walking me through your process

**Pull request merged**
